### PR TITLE
2.0 refactoring

### PR DESCRIPTION
**Pull request by andrew-codes**
- refactoring of core library
- introduction of high order functions (connectors) to provide sdk with AJAX implementations

### CODE REVIEW

File: src/createMeta.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        executeOperation(oidToken, operationName) {
+            invariant(oidToken, `Error: there was no \`oidToken\` provided to execute operation`);
+            invariant(operationName, `Error: there was no \`operationName\` provided to execute operation`);
+            const oid = new Oid(oidToken);
+            const url = `${urls.rest}/${oid.assetType}/${oid.number}?op=${operationName}`;
+            return postFn(url, null, headers);
+        },
+
+        queryDefinition(assetType) {
+            invariant(assetType, `Error: there was no \`assetType\` provided to queryDefinition`);
```
Comment (walkerrandolphsmith) [inline src/createMeta.js]: v1.meta does not require an assetType and if ignored will return the entire definition of Meta.

File: src/connectors/jqueryConnector.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+        method: 'POST',
+        data,
+        dataType: 'json',
+        headers,
+        success: resolve,
+        error: reject
+    })),
+    (url, data, headers) => new Promise((resolve, reject) => $.ajax(url, {
+        method: 'GET',
+        data: JSON.stringify(data),
```
Comment (JogoShugh) [inline src/connectors/jqueryConnector.js]: Does jquery not take care of query string Parameters or is that what it uses data for here?

File: src/createMeta.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            const url = `${urls.meta}/${assetType}`;
+            return getFn(url, null, headers);
+        }
+    };
+};
+
+const createHeaderObj = (token) => ({
+    Accept: 'application/json',
+    'Content-Type': 'application/json',
+    Authorization: `Basic ${token}`
```
Comment (JogoShugh) [inline src/createMeta.js]: Tokens are prefixed with Bearer not basic.

Comment (walkerrandolphsmith) [inline src/createMeta.js]: I hope there is not a hard requirement on providing an access token. Some consumers (spigot) need to use basic authentication to perform intentions by a specific Member.

Comment (andrew-codes) [inline src/createMeta.js]: @walkerrandolphsmith there is not, you can use an access token or username/password via `withCreds('username', 'password')`

File: src/connectors/jqueryConnector.js (hunk 2 above)
Comment (andrew-codes) [inline src/connectors/jqueryConnector.js]: @JogoShugh I believe jQuery handles the data as a query string with the method is `'GET'`

**Comment by andrew-codes** [comment]
@JogoShugh and @walkerrandolphsmith, I addressed the concerns you had about:
* authorization bearer versus basic
* querying meta definitions for all of meta instead of just for a single asset type

Any other concerns?

**Comment by walkerrandolphsmith** [comment]
Yes I have an issue with requiring an access token. Spigot requires actions be made on the behalf of specific users. Members can generate personal access tokens which could satisfy this case, however spigot is an automation tool and generating access tokens is not an automated process leaving spigot to require user interaction to work. If you allow basic authentication or access tokens then this will not be an issue for consumers like spigot. @andrew-codes

File: src/transformDataToAsset.js [hunk 4]
```diff
@@ -1,34 +1,42 @@
...
+const reduceAssetData = (obj) => Object.keys(obj)
+    .reduce((output, key) => {
         const attributeData = obj[key];
+        const assetAttributeKey = upperCamelCase(key);
         if (Array.isArray(attributeData)) {
-            output[key] = {
-                name: key,
-                value: attributeData.map(reduceRelationalAttributes)
+            return {
+                ...output,
```
Comment (JogoShugh) [inline src/transformDataToAsset.js]: Wow...so I'm assuming "output" is an object here, and ...output in this context means
expand all its properties into this new literal obejct?

**Pull request merged**